packages/aws-cdk/lib/logging.ts

```diff
@@ -108,105 +46,251 @@ export async function withCorkedLogging<T>(block: () => Promise<T>): Promise<T>
   } finally {
     CORK_COUNTER--;
     if (CORK_COUNTER === 0) {
-      logBuffer.forEach(([stream, str]) => stream.write(str + '\n'));
+      // Process each buffered message through notify
+      for (const ioMessage of logBuffer) {
+        void CliIoHost.getIoHost().notify(ioMessage);
+      }
       logBuffer.splice(0);
     }
   }
 }
 
+interface LogOptions {
+  /**
+   * The log level to use
+   */
+  level: IoMessageLevel;
+  /**
+   * The message to log
+   */
+  message: string;
+  /**
+   * Whether to force stdout
+   * @default false
+   */
+  forceStdout?: boolean;
+  /**
+ * Message code of the format [CATEGORY]_[NUMBER_CODE]
+ * @pattern [A-Z]+_[0-2][0-9]{3}
+ * @default TOOLKIT_[0/1/2]000
+ */
+  code?: string;
+}
+
 /**
- * Core logging function that handles all log output.
- * @param entry - LogEntry object or log level
- * @param fmt - Format string (when using with log level)
- * @param args - Format arguments (when using with log level)
+ * Internal core logging function that writes messages through the CLI IO host.
+ * @param options Configuration options for the log message. See  {@link LogOptions}
  */
-export function log(entry: LogEntry): void;
-export function log(level: LogLevel, fmt: string, ...args: unknown[]): void;
-export function log(levelOrEntry: LogLevel | LogEntry, fmt?: string, ...args: unknown[]): void {
-  // Normalize input
-  const entry: LogEntry = typeof levelOrEntry === 'string'
-    ? { level: levelOrEntry as LogLevel, message: util.format(fmt!, ...args) }
-    : levelOrEntry;
-
-  // Check if we should log this level
-  if (levelPriority[entry.level] > levelPriority[currentLogLevel]) {
+function log(options: LogOptions) {
+  if (levelPriority[options.level] > levelPriority[currentIoMessageThreshold]) {
     return;
   }
 
-  // Format the message
-  let finalMessage = entry.message;
-
-  // Add timestamp first if requested
-  if (entry.timestamp) {
-    finalMessage = `[${formatTime(new Date())}] ${finalMessage}`;
-  }
-
-  // Add prefix AFTER timestamp
-  if (entry.prefix) {
-    finalMessage = `${entry.prefix} ${finalMessage}`;
+  if (options.code) {
+    if (!validateMessageCode(options.code, options.level)) {
+      throw new Error(`Invalid message code: ${options.code}`);
+    }
+  } else {
+    options.code = `TOOLKIT_${Math.min(levelPriority[options.level] - 1, 2)}000`;
```

Comment:
did we decide on this way of leveling here? and i agree with codecov on this one, should be tested

Reply:
feels odd that we would have one system of 1-4, and another system of 0-2.

Reply:
bumping this